emily/lambda/src/main.rs

```diff
@@ -0,0 +1,178 @@
+            Ok(ApiGatewayProxyResponse {
+                status_code: 501,
+                multi_value_headers: headers.clone(),
+                is_base64_encoded: false,
+                body: Some(Body::Text(serialized_response)),
+                headers,
+            })
+        },
+    }
+}
+
+/// Main entry point for the AWS Lambda function.
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    // Run the lambda service.
+    lambda_runtime::run(
+        service_fn(
+            |event: LambdaEvent<ApiGatewayProxyRequest>| handle_event(event)
+        )
+    ).await
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+
+    fn create_event(method: Method, resource: &str, body: Option<String>) -> LambdaEvent<ApiGatewayProxyRequest> {
+        LambdaEvent {
+            payload: ApiGatewayProxyRequest {
+                http_method: method,
+                resource: Some(resource.to_string()),
+                body,
+                ..Default::default()
+            },
+            context: Default::default()
+        }
+    }
+
+    #[tokio::test]
+    async fn test_post_deposits_successful() {
+        // Arrange.
+        let json_body = serde_json::to_string(&CreateDepositRequestContent {
+            bitcoin_txid: "BITCOIN_TXID_UNITTEST".to_string(),
+            bitcoin_tx_output_index: 0.0,
+            reclaim: "RECLAIM_UNITTEST".to_string(),
+            deposit: "DEPOSIT_UNITTEST".to_string(),
+        }).unwrap();
+        let event = create_event(Method::POST, "/deposits", Some(json_body));
+
+        // Act.
+        let response = handle_event(event).await.expect("Failed to handle event");
+
+        // Assert.
+        assert_eq!(response.status_code, 201);
+        assert_eq!(response.body.unwrap(), Body::Text("{\"bitcoinTxid\":\"BITCOIN_TXID_UNITTEST\",\"bitcoinTxOutputIndex\":0.0,\"recipient\":\"MOCK_RECIPIENT\",\"amount\":11111.0,\"status\":\"PENDING\",\"statusMessage\":\"MOCK_CREATE_DEPOSIT_RESPONSE\",\"parameters\":{\"maxFee\":33333.0,\"lockTime\":22222.0,\"reclaimScript\":\"MOCK_RECLAIM_SCRIPT\"}}".to_string()));
+    }
+
+    #[test_case(Method::GET, "/deposits/{txid}/{outputIndex}", None; "get-deposit")]
+    #[test_case(Method::GET, "/deposits/{txid}", None; "get-txn-deposits")]
+    #[test_case(Method::GET, "/deposits", None; "get-deposits")]
+    #[test_case(Method::PUT, "/deposits", None; "update-deposits")]
+    #[test_case(Method::POST, "/withdrawals", None; "create-withdrawal")]
+    #[test_case(Method::GET, "/withdrawals/{id}", None; "get-withdrawal")]
+    #[test_case(Method::GET, "/withdrawals", None; "get-withdrawals")]
+    #[test_case(Method::PUT, "/withdrawals", None; "update-withdrawals")]
+    #[test_case(Method::GET, "/chainstate/{height}", None; "get-chainstate")]
+    #[test_case(Method::PUT, "/chainstate/{height}", None; "set-chainstate")]
+    #[test_case(Method::POST, "/chainstate/{height}", None; "update-chainstate")]
+    #[tokio::test]
```

Comment:
I'd rather have the `test_case` traits defined after `tokio::test` but unfortunately that's not possible. Not a big problem ~

- https://github.com/frondeus/test-case/pull/48